README.md
# k-icos

[![Latest Release](https://img.shields.io/github/v/tag/kalisio/k-icos?sort=semver&label=latest)](https://github.com/kalisio/k-icos/releases)
[![Build Status](https://app.travis-ci.com/kalisio/k-icos.svg?branch=master)](https://app.travis-ci.com/kalisio/k-icos)

A [Krawler](https://kalisio.github.io/krawler/) based service to download atmospheric data from the French gelectrical network operated by RTE.

## Description

The **k-icos** job allow to scrape atmospheric data from the [RTE API]. The downloaded data are stored within a [MongoDB](https://www.mongodb.com/) database and more precisely in 2 collections:
* the `icos-observations` collection stores the observation data
* the `icos-stations` collection stores the station positions

All records are stored in [GeoJson](https://fr.wikipedia.org/wiki/GeoJSON) format.

The project consists in 2 jobs:
* the `stations` job scrapes the available stations according a specific cron expression. By default, every day at midnight.
* the `observations` job scrapes the observations data according a specific cron expression. By default every hour.

## Configuration

### Stations

| Variable | Description |
|--- | --- |
| `DB_URL` | The database URL. The default value is `mongodb://127.0.0.1:27017/icos` |
| `OBJECT_SPEC_FILTER` | A filter used to select only latest data objects containing this pattern. The default value is `radon data`, could be eg `CO data`, `CO2 data`, `CH4 data` |
| `HISTORY` | The duration of the observations data history the job has to download. It must be expressed in seconds and the default value is `86 400 000` (1 day) | 
| `DEBUG` | Enables debug output. Set it to `krawler*` to enable full output. By default it is undefined. |

### Observations

| Variable | Description |
|--- | --- |
| `DB_URL` | The database URL. The default value is `mongodb://127.0.0.1:27017/icos` |
| `TTL` | The observations data time to live. It must be expressed in seconds and the default value is `604 800` (7 days) |
| `OBJECT_SPEC_FILTER` | A filter used to select only latest data objects containing this pattern. The default value is `radon data`, could be eg `CO data`, `CO2 data`, `CH4 data` |
| `OBJECT_VARIABLE` | The name of the variable contained in the corresponding data objects. The default value is `rn`, could be eg `co`, `co2`, `ch4` |
| `HISTORY` | The duration of the observations data history the job has to download. It must be expressed in seconds and the default value is `86 400 000` (1 day) | 
| `DEBUG` | Enables debug output. Set it to `krawler*` to enable full output. By default it is undefined. |

## Deployment

We personally use [Kargo](https://kalisio.github.io/kargo/) to deploy the service.

## Contributing

Please refer to [contribution section](./CONTRIBUTING.md) for more details.

## Authors

This project is sponsored by 

![Kalisio](https://s3.eu-central-1.amazonaws.com/kalisioscope/kalisio/kalisio-logo-black-256x84.png)
![IRSN](https://s3.eu-central-1.amazonaws.com/kalisioscope/assets/logos/irsn.png)

## License

This project is licensed under the MIT License - see the [license file](./LICENSE) for details

CHANGELOG.md
Please refer to GitHub issues for changelog: https://github.com/kalisio/k-icos/issues.


jobfile-stations.js
import _ from 'lodash'
import moment from 'moment'
import path from 'path'
import { fileURLToPath } from 'url'

const __dirname = path.dirname(fileURLToPath(import.meta.url))
const dbUrl = process.env.DB_URL || 'mongodb://127.0.0.1:27017/icos'
const specFilter = process.env.OBJECT_SPEC_FILTER || 'radon data'
const history =  +process.env.HISTORY || (1 * 24 * 60 * 60) // duration in seconds

export default {
  id: 'icos-stations',
  store: 'memory',
  options: {
    workersLimit: 1,
    faultTolerant: true,
  },
  tasks: [{
    id: 'stations',
    type: 'http',
    options: {
      url: 'https://meta.icos-cp.eu/sparql',
      method: 'POST',
      // This SPARQL request retrieve the latest data objects containing a specific label
      // Then it explore relationships to retrieve the associated station information
      body: `prefix cpmeta: <http://meta.icos-cp.eu/ontologies/cpmeta/>
            prefix prov: <http://www.w3.org/ns/prov#>
            prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#>

            select (str(?submTime) as ?time) ?object ?spec ?fileName ?station ?stationId ?stationName ?latitude ?longitude ?altitude where{
              ?object cpmeta:hasObjectSpec ?specUri .
              ?object cpmeta:wasSubmittedBy/prov:endedAtTime ?submTime .
              ?object cpmeta:hasName ?fileName .
              ?object cpmeta:wasAcquiredBy ?acquisition .
              ?acquisition prov:wasAssociatedWith ?station .
              ?station cpmeta:hasLatitude ?latitude .
              ?station cpmeta:hasLongitude ?longitude .
              ?station cpmeta:hasElevation ?altitude .
              ?station cpmeta:hasStationId ?stationId .
              ?station cpmeta:hasName ?stationName .
              ?specUri rdfs:label ?spec .
              filter contains(?spec, "${specFilter}")
              filter (?submTime > "${moment.utc().subtract(history, 'seconds').format()}"^^xsd:dateTime)
            }
            order by desc(?submTime)`,
      type: 'JSON'
    }
  }],
  hooks: {
    tasks: {
      after: {
        readJson: {
          objectPath: 'results.bindings',
          transform: {
            mapping: {
              'longitude.value': 'longitude',
              'latitude.value': 'latitude',
              'altitude.value': 'altitude',
              'stationId.value': 'stationId',
              'stationName.value': 'stationName'
            },
            pick: ['longitude', 'latitude', 'altitude', 'stationId', 'stationName']
          }
        },
        convertToGeoJson: {
        },
        updateMongoCollection: {
          collection: 'icos-stations',
          filter: { 'properties.stationId': '<%= properties.stationId %>' },
          upsert: true,
          chunkSize: 256
        },
        clearData: {}
      }
    },
    jobs: {
      before: {
        createStores: [{
          id: 'memory'
        }, {
          id: 'fs', options: { path: __dirname }
        }],
        connectMongo: {
          url: dbUrl,
          // Required so that client is forwarded from job to tasks
          clientPath: 'taskTemplate.client'
        },
        createMongoCollection: {
          clientPath: 'taskTemplate.client',
          collection: 'icos-stations',
          indices: [
            [{ 'properties.stationId': 1 }, { unique: true }], 
            { geometry: '2dsphere' }
          ]
        }
      },
      after: {
        disconnectMongo: {
          clientPath: 'taskTemplate.client'
        },
        removeStores: [ 'memory', 'fs' ]
      },
      error: {
        disconnectMongo: {
          clientPath: 'taskTemplate.client'
        },
        removeStores: [ 'memory', 'fs' ]
      }
    }
  }
}


jobfile-observations.js
import _ from 'lodash'
import moment from 'moment'
import path from 'path'
import { fileURLToPath } from 'url'

const __dirname = path.dirname(fileURLToPath(import.meta.url))

const dbUrl = process.env.DB_URL || 'mongodb://127.0.0.1:27017/icos'
const ttl = +process.env.TTL || (7 * 24 * 60 * 60)  // duration in seconds
const history =  parseInt(process.env.HISTORY) || (1 * 24 * 60 * 60) // duration in seconds
const specFilter = process.env.OBJECT_SPEC_FILTER || 'radon data'
const variable = process.env.OBJECT_VARIABLE || 'rn'
const startTime = moment.utc().subtract(history, 'seconds')

export default {
  id: 'icos-observations',
  store: 'fs',
  options: {
    workersLimit: 1
  },
  taskTemplate: {
    // Default file name is raw data (eg as zip) instead of CSV, update to avois any confusion
    id: `<%= fileName.replace('.zip', '.csv') %>`,
    type: 'http',
    options: {
      // URL to download objects directly as CSV is slightly different from raw data (eg as zip)
      // See https://github.com/ICOS-Carbon-Portal/data#csv-download-for-tabular-time-series-data
      url: `<%= objectUrl.replace('meta.icos-cp.eu/objects', 'data.icos-cp.eu/csv') %>`,
      auth: {
        url: 'https://cpauth.icos-cp.eu/password/login',
        form: {
          mail: process.env.USER_EMAIL,
          password: process.env.USER_PASSWORD
        }
      }
    }
  },
  hooks: {
    tasks: {
      before: {
        createMongoAggregation: {
          dataPath: 'data.mostRecentData',
          collection: 'icos-observations',
          pipeline: [
            {
              $match: {
                'properties.samplingHeight': '<%= samplingHeight %>',
                'properties.stationId': '<%= stationId %>',
                [`properties.${variable}`]: { $exists: true }
              }
            },
            { $sort: { time: -1 } },
            { $limit : 1 },
            {
              $group: {
                _id: "$properties.stationId",
                time: { $first: "$time" }
              }
            }
          ],
          allowDiskUse: true
        }
      },
      after: {
        readCSV: {
          headers: true
        },
        apply: {
          function: (item) => {
            const { longitude, latitude, altitude, samplingHeight, stationId, stationName, data } = item
            const latestData = _.get(item, 'mostRecentData[0]')
            if (latestData) console.log('Found previous observation data for station ' + stationId + ' at ' + samplingHeight + 'm')
            let features = []
            _.forEach(data, (record) => {
              const time = moment.utc(record.TIMESTAMP)
              // Check if newer
              if (latestData && time.isSameOrBefore(moment.utc(latestData.time))) return
              // If so push it
              features.push({
                time: time.toDate(),
                level: samplingHeight,
                type: 'Feature',
                geometry: {
                  type: 'Point',
                  coordinates: [longitude, latitude, altitude + samplingHeight]
                },
                properties: {
                  stationId,
                  stationName,
                  samplingHeight,
                  [variable]: _.toNumber(record[variable])
                }
              })
            })
            if (features.length > 0) console.log('Found ' + features.length + ' new observation data for station ' + stationId + ' at ' + samplingHeight + 'm')
            else console.log('No new observation data found for station ' + stationId + ' at ' + samplingHeight + 'm')
            item.data = features
          }
        },
        writeMongoCollection: {
          collection: 'icos-observations',
          chunkSize: 256
        },
        clearOutputs: {}
      }
    },
    jobs: {
      before: {
        basicAuth: { jar: true, optionsPath: 'taskTemplate.options' }, // Set cookie
        createStores: [{
          id: 'memory'
        }, {
          id: 'fs', options: { path: __dirname }
        }],
        connectMongo: {
          url: dbUrl,
          // Required so that client is forwarded from job to tasks
          clientPath: 'taskTemplate.client'
        },
        createMongoCollection: {
          clientPath: 'taskTemplate.client',
          collection: 'icos-observations',
          indices: [
            { 'properties.stationId': 1 },
            { [`properties.${variable}`]: 1 },
            { 'properties.stationId': 1, time: -1 },
            { 'properties.stationId': 1, [`properties.${variable}`]: 1, time: -1 },
            [{ time: 1 }, { expireAfterSeconds: ttl }], // days in s
            { geometry: '2dsphere' }                                                                                                              
          ],
        },
        runTask: {
          store: 'memory',
          id: 'data-objects',
          type: 'http',
          options: {
            url: 'https://meta.icos-cp.eu/sparql',
            method: 'POST',
            // This SPARQL request retrieve the latest data objects containing a specific label
            // Then it explore relationships to retrieve the associated station information
            body: `prefix cpmeta: <http://meta.icos-cp.eu/ontologies/cpmeta/>
                  prefix prov: <http://www.w3.org/ns/prov#>
                  prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#>

                  select (str(?submTime) as ?time) ?object ?spec ?fileName ?station ?samplingHeight ?stationId ?stationName ?latitude ?longitude ?altitude where{
                    ?object cpmeta:hasObjectSpec ?specUri .
                    ?object cpmeta:wasSubmittedBy/prov:endedAtTime ?submTime .
                    ?object cpmeta:hasName ?fileName .
                    ?object cpmeta:wasAcquiredBy ?acquisition .
                    ?acquisition prov:wasAssociatedWith ?station .
                    ?acquisition cpmeta:hasSamplingHeight ?samplingHeight .
                    ?station cpmeta:hasLatitude ?latitude .
                    ?station cpmeta:hasLongitude ?longitude .
                    ?station cpmeta:hasElevation ?altitude .
                    ?station cpmeta:hasStationId ?stationId .
                    ?station cpmeta:hasName ?stationName .
                    ?specUri rdfs:label ?spec .
                    filter contains(?spec, "${specFilter}")
                    filter (?submTime > "${moment.utc().subtract(history, 'seconds').format()}"^^xsd:dateTime)
                  }
                  order by desc(?submTime)`,
            type: 'JSON'
          }
        },
        readJson: {
          store: 'memory',
          key: 'data-objects',
          objectPath: 'results.bindings',
          transform: {
            mapping: {
              'object.value': 'objectUrl',
              'fileName.value': 'fileName',
              'longitude.value': 'longitude',
              'latitude.value': 'latitude',
              'altitude.value': 'altitude',
              'samplingHeight.value': 'samplingHeight',
              'stationId.value': 'stationId',
              'stationName.value': 'stationName'
            },
            unitMapping: {
              longitude: { asNumber: true },
              latitude: { asNumber: true },
              altitude: { asNumber: true },
              samplingHeight: { asNumber: true }
            },
            pick: ['objectUrl', 'fileName', 'longitude', 'latitude', 'altitude', 'samplingHeight', 'stationId', 'stationName']
          },
          // Generate a download task per data object
          dataPath: 'data.tasks'
        }
      },
      after: {
        disconnectMongo: {
          clientPath: 'taskTemplate.client'
        },
        removeStores: [ 'memory', 'fs' ]
      },
      error: {
        disconnectMongo: {
          clientPath: 'taskTemplate.client'
        },
        removeStores: [ 'memory', 'fs' ]
      }
    }
  }
}
